umbral/umbral.py

```diff
@@ -115,98 +116,135 @@ def __bytes__(self):
+                Passing both is also fine.")
+
+        self._point_eph_e = point_eph_e
+        self._point_eph_v = point_eph_v
+        self._bn_sig = bn_sig
+
+        self._point_eph_e_prime = e_prime
+        self._point_eph_v_prime = v_prime
+        self._point_noninteractive = noninteractive_point
 
         self.cfrags = {}
+        self._contents = None
 
-    @staticmethod
-    def from_bytes(data: bytes, curve: ec.EllipticCurve):
+    @classmethod
+    def from_original_bytes(cls, data: bytes, curve: ec.EllipticCurve):
         """
         Instantiates a Capsule object from the serialized data.
         """
         eph_e = Point.from_bytes(data[0:33], curve)
         eph_v = Point.from_bytes(data[33:66], curve)
         sig = BigNum.from_bytes(data[66:98], curve)
 
-        return Capsule(eph_e, eph_v, sig)
+        return cls(eph_e, eph_v, sig)
+
+    @classmethod
+    def from_reconstructed_bytes(cls, data: bytes, curve: ec.EllipticCurve):
+        """
+        Instantiate a Capsule from serialized data after reconstruction has occurred.
+
+        The most obvious use case is Bob affixing at least m cFrags and then serializing the Capsule.
+        """
+        # TODO: Seems like a job for BytestringSplitter ?
+        e_prime = Point.from_bytes(data[0:33], curve)
+        v_prime = Point.from_bytes(data[33:66], curve)
+        eph_ni = Point.from_bytes(data[66:99], curve)
+
+        return cls(e_prime=e_prime, v_prime=v_prime, noninteractive_point=eph_ni)
+
+    @property
+    def contents(self):
+        return self._contents
 
     def to_bytes(self):
         """
         Serialize the Capsule into a bytestring.
         """
-        eph_e = self.point_eph_e.to_bytes()
-        eph_v = self.point_eph_v.to_bytes()
-        sig = self.bn_sig.to_bytes()
+        eph_e = self._point_eph_e.to_bytes()
+        eph_v = self._point_eph_v.to_bytes()
+        sig = self._bn_sig.to_bytes()
 
         return eph_e + eph_v + sig
 
     def verify(self, params: UmbralParameters):
 
-        e = self.point_eph_e
-        v = self.point_eph_v
-        s = self.bn_sig
+        e = self._point_eph_e
+        v = self._point_eph_v
+        s = self._bn_sig
         h = hash_to_bn([e, v], params)
 
         return params.g * s == v + (e * h)
 
     def attach_cfrag(self, cfrag: CapsuleFrag):
         self.cfrags[cfrag.bn_kfrag_id] = cfrag
 
-    def reconstruct(self):
+    def original_ephemeral_points(self):
+        return self._point_eph_e + self._point_eph_v
+
+    def reconstructed_ephemeral_points(self):
+        return self._point_eph_e_prime + self._point_eph_v_prime
+
+    def open(self, pub_bob, priv_bob, pub_alice, force_reopen=False, pre=None):
+        # TODO: Raise an error here if Bob has gathered enough cFrags.
+        if self.contents and not force_reopen:
+            newly_opened = True
+        else:
+            self._reconstruct(pre=pre)
+            if not pre:
+                pre = PRE(UmbralParameters())
```

Comment:
UmbralParameters() should not be called here. There is only ONE set of parameters for the whole system, and UmbralParameters creates these parameters, where some of them are randomized. Therefore if we create new parameters, things will start not to work. Perhaps I should have used some kind of singleton pattern for this parameters, of something similar.

Reply:
@cygnusv absolutely - I was going to make an issue for this next.

Let's turn the PRE class into a module, and instantiate the one and only `UmbralParamteres` there.

Consider this a workaround until we do that.